src/extensions/nodejsMigrationHelper.ts

```diff
@@ -0,0 +1,57 @@
+import { FirebaseError } from "../error";
+import * as extensionsApi from "./extensionsApi";
+import { logPrefix } from "./extensionsHelper";
+import { promptOnce } from "../prompt";
+import * as utils from "../utils";
+
+const billingMsgUpdate =
+  "This update includes an upgrade to Node.js 10 from Node.js 8, which is no longer maintained. Node.js 10 requires your project to be on the Blaze (pay as you go) plan, and starting with this update, you will be charged a small amount (typically around $0.01/month) for each instance of this extension you've installed, in addition to any charges associated with your use of Firebase services.";
+const billingMsgCreate =
+  "This extension includes Node.js 10 functions. Node.js 10 requires your project to be on the Blaze (pay as you go) plan, and you will be charged a small amount (typically around $0.01/month) for each instance of this extension you've installed, in addition to any charges associated with your use of Firebase services.";
+
+function hasRuntime(resource: extensionsApi.Resource, runtime: string): boolean {
+  let resourceRuntime = "nodejs8"; // default to nodejs8 runtime
+  if (resource.properties?.runtime) {
+    resourceRuntime = resource.properties.runtime;
+  }
+  return resourceRuntime == runtime;
+}
+
+function shouldDisplayMsg(
+  newSpec: extensionsApi.ExtensionSpec,
+  curSpec?: extensionsApi.ExtensionSpec
+): boolean {
+  const newResources = newSpec.resources;
```

Comment:
is there a reason why you didn't write this as:
`const newResources = newSpec.resources || [];`

Reply:
I assumed that resources would never be empty, but that now realized that isn't a good assumption. Fixed.